### ISSUE TITLE
Preview for Intermediates in dagit

### ISSUE THREAD

**Issue by newturok** [dagster-io/dagster#2630]
> This can be very useful for debugging, when you want to inspect the value output by a solid and ensure that it is as you expect; for audit, when you want to understand how a particular downstream output was created; and for re-executing downstream solids with cached results from expensive upstream computations.

Would be perfect, if you can preview text representations for ```Intermediates``` inside ```dagit```
Somehow like this
![image](https://user-images.githubusercontent.com/5757827/85210712-0372dd80-b34b-11ea-854c-196b9109ed7d.png)

**Comment by deem0n** [comment]
I would vote for inline value printing, exactly like we print `key` for intermediate. It is a killer feature for real life usage!
something like this
![image](https://user-images.githubusercontent.com/871385/103104832-b3970b00-463a-11eb-91bd-6301e0e8d89b.png)

**Comment by schrockn** [comment]
So early on in the project we had this type of capability added by default. However, this was unacceptable for situations where users were dealing with sensitive information.

In lieu of that you can implement your own custom type to do this, and yield the appropriate EventMetadataEntry.

For example, here is an example of adding a "preview" row to a pandas data frame type. (Note: not particularly proud of the call into the protected method) The important bit is the extra call to create the markdown metadata entry.

If you want it totally inline the value you can just use EventMetadataEntry.text

```
from dagster import DagsterType, EventMetadataEntry, TypeCheck
from dagster_pandas import DataFrame as PandasDataFrame


def type_check_with_metadata(context, value):
    res = PandasDataFrame._type_check_fn(context, value)  # pylint: disable=protected-access
    if not res.success:
        return res

    return TypeCheck(
        res.success,
        res.description,
        res.metadata_entries
        + [
            EventMetadataEntry.md(
                md_str=value.head().to_markdown(), label="df_head", description="Head of data frame"
            )
        ],
    )


DataFrame = DagsterType(
    name="PandasDataFrame",
    description=(
        "Two-dimensional size-mutable, potentially heterogeneous tabular data structure with "
        "labeled axes (rows and columns). See http://pandas.pydata.org/"
    ),
    loader=PandasDataFrame.loader,
    materializer=PandasDataFrame.materializer,
    type_check_fn=type_check_with_metadata,
)
```

**Comment by catherinewu** [comment]
cc @yuhan to consider including in advanced usage guides

**Comment by yuhan** [comment]
@catherinewu i'm hesitant to write up a guide which uses protected methods. IIRC, the asset lane is working on something that can potentially address this use case.
cc @OwenKephart is this something the new "structured metadata" can tackle?

**Comment by schrockn** [comment]
We should definitely not add protected method usages to public usage guides :-) I had just copied and pasted that from working code

**Comment by OwenKephart** [comment]
@yuhan The changes will allow you to attach EventMetadataEntries to Outputs (which will end up on the STEP_OUTPUT event) or yield them from IOManagers (which will end up on the HANDLED_OUTPUT event). Those metadata entries will show up in the event log in the way you would expect. We don't have anything specifically for intermediates, but as far as I know those will be deprecated in favor of IOManagers regardless (and the conversion should be fairly painless).

TLDR; yep our changes should handle this issue

**Comment by yuhan** [comment]
Tracking content work in https://github.com/dagster-io/dagster/issues/3774

**Comment by sryza** [comment]
Closing this given that we can now approximate this functionality and the content work is tracked elsewhere.
